### ISSUE TITLE
Ping excessive delay or loop halt when target offline

### ISSUE THREAD

**Issue by drakerex** [arduino-libraries/WiFiNINA#18]
As requested from https://forum.arduino.cc/index.php?topic=569872.0

When a WiFi.ping is sent to a host (by dns) which is offline, there's an excessive delay. If the WiFi.ping command is sent by IPAddress the loop() halts. See the attached forum for example sketch.

**Comment by sandeepmistry** [comment]
Hi @drakerex,

I tried to reproduce this now.

1) The bad DNS stall added 10s, so I've submitted pull request #21 to resolve this.

2) I could not reproduce the "halt" however, the firmware has a 5 second timeout for the ping result. Is this what you are seeing?

**Comment by apetryk2** [comment]
I can reproduce the 'halt' issue with the following code.  The IP address I'm trying to ping (192.168.0.2 in this case) doesn't belong to anything.  The debug trace gets to "calling ping" and then just hangs.

`
#include <SPI.h>
#include <WiFiNINA.h>
#include <utility/wifi_drv.h>

///////please enter your sensitive data in the Secret tab/arduino_secrets.h
char ssid[] = SECRET_SSID;        // your network SSID (name)
char pass[] = SECRET_PASS;    // your network password (use for WPA, or use as key for WEP)
int status = WL_IDLE_STATUS;     // the WiFi radio's status
char debug = 1;

// Specify IP address or hostname
IPAddress pingHost(192,168,0,2);
int pingResult;

void setup() {
  
  if (debug) {
    //Initialize serial and wait for port to open:
    Serial.begin(9600);
    while (!Serial) {
      ; // wait for serial port to connect. Needed for native USB port only
    }

    // check for the WiFi module:
    if (WiFi.status() == WL_NO_MODULE) {
      Serial.println("Communication with WiFi module failed!");
      // don't continue
      while (true);
    }
  
    String fv = WiFi.firmwareVersion();
    if (fv < WIFI_FIRMWARE_LATEST_VERSION) {
      Serial.println("Please upgrade the firmware");
    }
  }
}

void loop() {

  TestWiFiConnection();

  if (debug) Serial.println("calling ping");
  pingResult = WiFi.ping(pingHost);
  if (debug) Serial.println("done calling ping");

  delay(500);
}

void TestWiFiConnection()
//test if always connected
{
  int StatusWiFi=WiFi.status();
  if(StatusWiFi!=WL_CONNECTED) //if no connection
  {
    WiFiConnect(); //if my SSID is present, connect
  }
}

void WiFiConnect()
//connect to my SSID
{
  status= WL_IDLE_STATUS;
  while(status!=WL_CONNECTED)
  {
    status = WiFi.begin(ssid,pass);
    if(status!=WL_CONNECTED) delay(500);
  }
  if (debug) printWiFiData();
}

void printWiFiData() {
  // print your board's IP address:
  IPAddress ip = WiFi.localIP();
  Serial.print("IP address : ");
  Serial.println(ip);

  Serial.print("Subnet mask: ");
  Serial.println((IPAddress)WiFi.subnetMask());

  Serial.print("Gateway IP : ");
  Serial.println((IPAddress)WiFi.gatewayIP());

  // print your MAC address:
  byte mac[6];
  WiFi.macAddress(mac);
  Serial.print("MAC address: ");
  printMacAddress(mac);
}

void printCurrentNet() {
  // print the SSID of the network you're attached to:
  Serial.print("SSID: ");
  Serial.println(WiFi.SSID());

  // print the MAC address of the router you're attached to:
  byte bssid[6];
  WiFi.BSSID(bssid);
  Serial.print("BSSID: ");
  printMacAddress(bssid);
  // print the received signal strength:
  long rssi = WiFi.RSSI();
  Serial.print("signal strength (RSSI): ");
  Serial.println(rssi);

  // print the encryption type:
  byte encryption = WiFi.encryptionType();
  Serial.print("Encryption Type: ");
  Serial.println(encryption, HEX);
  Serial.println();
}

void printMacAddress(byte mac[]) {
  for (int i = 5; i >= 0; i--) {
    if (mac[i] < 16) {
      Serial.print("0");
    }
    Serial.print(mac[i], HEX);
    if (i > 0) {
      Serial.print(":");
    }
  }
  Serial.println();
}
`

**Comment by szaiftamas** [comment]
I changed a bit the spi_drv.cpp in the WiFiNINA library. I added timeout for these two function, and the hang is over, now I able to reconnect to the wifi router, if the connection is drop.
This problem happened when the wifi channel was overloaded, it means 20 wifi router in the same place, and the other when the signal strength can reached only the minimum level, and it is sometime lost the connection.

void SpiDrv::waitForSlaveSign()
{
	uint32_t timeOut = millis() + 10000;
	while (!waitSlaveSign() && timeOut > millis());
}

void SpiDrv::waitForSlaveReady()
{
    uint32_t timeOut = millis() + 10000;
	while (!waitSlaveReady() && timeOut > millis());
}
